File: day06/lights2.py
#!/usr/bin/env python
import fileinput
import re
import sys
import string
from pylab import *

XMAX = 1000
YMAX = 1000

# initialize
light = zeros([XMAX, YMAX])

# parse the instructions
for line in fileinput.input() :
    line = line.rstrip("\n")
    m = re.search('(\d+),(\d+) through (\d+),(\d+)', line)
    if m != None :
        (x1, y1, x2, y2) = m.group(1, 2, 3, 4)
        x1 = int(x1)
        y1 = int(y1)
        x2 = int(x2) + 1
        y2 = int(y2) + 1

    if string.find(line, "turn on") == 0:
        light[y1:y2, x1:x2] += 1

    if string.find(line, "turn off") == 0 :
        r = light[y1:y2, x1:x2]
        r -= 1
        r[r < 0] = 0
        light[y1:y2, x1:x2] = r
        

    if string.find(line, "toggle") == 0:
        light[y1:y2, x1:x2] += 2
        
    if string.find(line, "print") == 0:
        print light

total_brightness = sum(light)
print "total brightness = %d" % total_brightness

File: day06/lights2.pl
#!/usr/bin/env perl
use strict;
use warnings;
use 5.10.0;

my ($XMAX, $YMAX);

if (@ARGV >= 2) {
    $XMAX = shift @ARGV;
    $YMAX = shift @ARGV;
} else {
    $XMAX = $YMAX = 1000;
}

my @light;

# initialize
for my $x (0..$XMAX - 1) {
    for my $y (0..$YMAX - 1) {
        $light[$x][$y] = 0;
    }
}

# parse the instructions
while (<>) {
    chomp;
    m/(\d+),(\d+) through (\d+),(\d+)/;
    my ($x1, $y1, $x2, $y2) = ($1, $2, $3, $4);

    if (/^turn on/) {
        for my $x ($x1..$x2) {
            for my $y ($y1..$y2) {
                $light[$x][$y]++;
            }
        }
    }

    if (/^turn off/) {
        for my $x ($x1..$x2) {
            for my $y ($y1..$y2) {
                if ($light[$x][$y] > 0) {
                    $light[$x][$y]--;
                }
            }
        }
    }

    if (/^toggle/) {
        for my $x ($x1..$x2) {
            for my $y ($y1..$y2) {
                $light[$x][$y] += 2;
            }
        }
    }

    if (/^print/) {
        for my $x (0..$XMAX - 1) {
            for my $y (0..$YMAX - 1) {
                print "$light[$x][$y] ";
            }
            print "\n";
        }
    }
}

# what's the total brightness?
my $total = 0;
for my $x (0..$XMAX - 1) {
    for my $y (0..$YMAX - 1) {
        $total += $light[$x][$y];
    }
}

say "total brightness = $total";
